main.go

```diff
@@ -0,0 +1,436 @@
+						updateCache()
+						definitionErrors.Inc()
+						continue
+					}
+					log.Print(err)
+					//applyError(item, err)
+					recorder.Event(&item, corev1.EventTypeWarning, err.Error(), "")
+				}
+				continue
+			}
+
+			for id, cached := range alarmDefinitionCache {
+				// if exists, check if needs update
+				if item.Spec.ID == id && !equal(&item, cached) {
+					//update if possible
+					log.Printf("Updating alarm %s", item.Spec.Name)
+					err := updateAlarmDefinition(&item, defClient)
+					if err != nil {
+						log.Print(err)
+						definitionErrors.Inc()
+						recorder.Event(&item, corev1.EventTypeWarning, err.Error(), "")
+					}
+					continue discoveredLoop
+				}
+			}
+
+		}
+	}
+}
+
+func updateCache() error {
+	existing, err := monascaclient.GetAlarmDefinitions(nil)
+	if err != nil {
+		log.Print(err)
+		return err
+	}
+	//clear map
+	for k := range alarmDefinitionCache {
+		delete(alarmDefinitionCache, k)
+	}
+	for _, item := range existing.Elements {
+		//ignore all alarm definitions that do not have the adc suffix
+		if strings.HasSuffix(item.Name, alarmDefinitionControllerSuffix) {
+			alarmDefinitionCache[item.ID] = item
+		}
+	}
+
+	return nil
+}
+
+func setKeystoneToken() error {
+	opts, err := openstack.AuthOptionsFromEnv()
+	if err != nil {
+		log.Printf("Error obtaining auth settings from env: %s", err.Error())
+		return err
+	}
+
+	openstackProvider, err := openstack.AuthenticatedClient(opts)
+	if err != nil {
+		log.Printf("Error authenticating with keystone: %s", err.Error())
+		return err
+	}
+	//log.Print(openstackProvider.TokenID)
+	token := openstackProvider.TokenID
+	headers := http.Header{}
+	headers.Add("X-Auth-Token", token)
+	monascaclient.SetHeaders(headers)
+	return nil
+}
+
+func addAlarmDefinition(namespace string, r *v1.AlarmDefinition, sampleclientset clientset.Interface) error {
+	if *defaultNotification != "" && len(r.Spec.AlarmActions) <= 0 {
+		if defaultNotificationID == "" {
+			return fmt.Errorf("Unable to apply default notification method: no ID found")
+		}
+		r.Spec.AlarmActions = []string{defaultNotificationID}
+	}
+	definitionRequest := convertToADRequest(r)
+	glog.Warning(definitionRequest)
+	result, err := monascaclient.CreateAlarmDefinition(definitionRequest)
+	if err != nil {
+		return err
+	}
+	alarmDefinitionCache[result.ID] = *result
+	newDef := convertToADResource(r, result)
+	sampleclientset.Monasca().AlarmDefinitions(namespace).Update(newDef)
+
+	log.Printf("Added definition %v", r.Spec)
+	return nil
+}
+
+func updateAlarmDefinition(r *v1.AlarmDefinition, sampleclientset clientset.Interface) error {
+	definitionRequest := convertToADRequest(r)
+	result, err := monascaclient.PatchAlarmDefinition(r.Spec.ID, definitionRequest)
+	if err != nil {
+		return err
+	}
+	alarmDefinitionCache[result.ID] = *result
+	newDef := convertToADResource(r, result)
+	sampleclientset.Monasca().AlarmDefinitions(*namespace).Update(newDef)
```

Comment:
Why don't you check for errors in some of your monasca client calls?

Reply:
this line is a kubernetes CRD client call, but I should check the error